go-controller/pkg/node/healthcheck.go

```diff
@@ -93,29 +96,120 @@ func countLocalEndpoints(ep *kapi.Endpoints, nodeName string) int {
+	for _, val := range values {
+		klog.Warningf("Found stale interface %s, so deleting it", val)
+		_, stderr, err := util.RunOVSVsctl("--if-exists", "--with-iface", "del-port", val)
+		if err != nil {
+			klog.Errorf("Failed to delete OVS port/interface %s: stderr: %s (%v)",
+				val, stderr, err)
+		}
+	}
+}
+
+// checkForStaleOVSRepresentorInterfaces checks for stale OVS ports backed by Repreresentor interfaces,
+// derrive iface-id from pod name and namespace then remove any interfaces assoicated with a sandbox that are
+// not scheduled to the node.
+func checkForStaleOVSRepresentorInterfaces(nodeName string, k kube.Interface) {
+	// list Ports on br-int and find out the current ports that have iface-ids
+	currentIfaceIds := make(map[string]string)
+	ports, stderr, err := util.RunOVSVsctl("list-ports", "br-int")
+	if err != nil {
+		klog.Errorf("failed to get list of ports on br-int:, stderr: %q, error: %v", stderr, err)
+		return
+	}
+
+	for _, port := range strings.Split(ports, "\n") {
+		stdout, stderr, err := util.RunOVSVsctl("get", "Port", port, "Interfaces")
+		if err != nil {
+			klog.Errorf("failed to get port %q on br-int:, stderr: %q, error: %v", port, stderr, err)
+			continue
+		}
+		// remove brackets on list of interfaces
+		ifaces := strings.TrimPrefix(strings.TrimSuffix(stdout, "]"), "[")
+		for _, iface := range strings.Split(ifaces, ",") {
+			// check that interface is associated with a sandbox
+			sandbox, stderr, err := util.RunOVSVsctl("get", "Interface", strings.TrimSpace(iface),
+				"external_ids:sandbox")
 			if err != nil {
-				klog.Errorf("Failed to list OVS interfaces with ofport set to -1")
+				if !strings.Contains(stderr, "no key") {
+					klog.Errorf("failed to get Interface %q external_ids:sandbox. stderr: %q, error: %v",
+						iface, stderr, err)
+				}
 				continue
 			}
-			if len(stdout) == 0 {
+			if sandbox == "" {
 				continue
 			}
-			values := strings.Split(stdout, "\n\n")
-			for _, val := range values {
-				klog.Warningf("Found stale interface %s, so deleting it", val)
-				_, stderr, err := util.RunOVSVsctl("--if-exists", "--with-iface", "del-port", val)
-				if err != nil {
-					klog.Errorf("Failed to delete OVS port/interface %s: stderr: %s (%v)",
-						val, stderr, err)
-				}
+			// get iface ID
+			ifaceId, stderr, err := util.RunOVSVsctl("get", "Interface", strings.TrimSpace(iface),
+				"external_ids:iface-id")
+			if err != nil {
+				klog.Errorf("failed to get Interface %q external_ids:iface-id. stderr: %q, error: %v",
+					iface, stderr, err)
+				continue
+			}
+			if ifaceId == "" {
+				continue
+			}
+			currentIfaceIds[ifaceId] = port
+		}
+	}
+
+	if len(currentIfaceIds) == 0 {
+		return
+	}
+
+	// list Pods and calculate the expected iface-ids.
+	// Note: we do this after scanning ovs ports to avoid deleting ports of pods that where just scheduled
+	// on the node.
+	pods, err := k.GetPods("", metav1.LabelSelector{}, fields.OneTermEqualSelector(
```

Comment:
The pod watcher on the node is already filtered for just pods on its node. Is there no feasible way to leverage that instead?

Reply:
The intention is to periodically clean up stale representor ports from OVS (if any) much like `checkForStaleOVSInterfaces` is doing today for internal ports.

To do so, we care about the current state of pods on this node at the point this periodic runs.
doing it with pod watcher would mean we add a new watcher that will update a newly defined async data struct as pods are created/deleted (e.g a map/list + mutex) and this periodic would then access this data struct to get the current pods allocated to the node.

Is this extra complexity needed ? did you have something else in mind ?

Reply:
@trozet any thoughts on this ?

Reply:
I think you can just use the node watch factory to get the pods from the informer cache, rather than issue kube client commands:

```
n.watchFactory.(*factory.WatchFactory).GetPods()
```

you may need to add a "GetLocalPods" function here:
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/factory/types.go#L13

Reply:
OK, modified the impl to use the node's WatchFactory (ObjectCacheInterface to be precise)
this in most cases will have only locally scheduled pods (unless [masterWatchFactory is used during node init](https://github.com/ovn-org/ovn-kubernetes/blob/7bd21ac3a82224c03f60aa65b4516265fde1a045/go-controller/cmd/ovnkube/ovnkube.go#L260))
in that case we would still need to filter pods by `spec.nodeName` but no k8s API call to get pods.